### DISCUSSION TITLE
How to configure  capacityProviderStrategy?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#18203] [Q&A]
I have a hybrid, ECS deployment that successfully runs Fargate tasks. I now want to extend it such that I can run some tasks on EC2 by setting the job's capacity provider. I created a new EC2 Auto-scaling Group on AWS and added it as a capacity provider to my Dagster ECS cluster. When I run a job with the `capacityProviderStrategy` tag set to my auto-scaling group I receive the following error immediately:
```botocore.errorfactory.InvalidParameterException: An error occurred (InvalidParameterException) when calling the RunTask operation: Assign public IP is not supported for this launch type.```
Any help is greatly appreciated.

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C01U954MEER/p1700599877859309?thread_ts=1700599877.859309&cid=C01U954MEER).*

**Comment by gibsondan** [comment]
Can you include the exact tags on the job that you used? 

You may also need to override one or both of the launchType and networkConfiguration fields as well in the same place that you set capacityProviderStrategy. (See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ecs/client/run_task.html#ECS.Client.run_task for more information on those fields)

The default networkConfiguration field is set here: https://github.com/dagster-io/dagster-cloud/blob/main/dagster-cloud/dagster_cloud/workspace/ecs/client.py#L109-L118 - it looks like it is smart enough to exclude assignPublicIp if the launchType is globally set to something other than FARGATE, but is not smart enough to set it if that launch type is later overridden via a tag.

**Comment by frs171717** [comment]
Thanks for quick reply, @gibsondan. Here is my current tags. I'm happy to try overriding those fields, mind showing exactly what you'd suggest?
```
 tags = {
    "ecs/run_task_kwargs": {
      "capacityProviderStrategy": [
        {
          "capacityProvider": "name-of-my-capacity-provider-as-seen-in-infrastructure-tab",
        },
      ],
    },
  }
```

**Reply by gibsondan** [reply]
The proposal is to add a "launchType": "EC2" key and a "networkConfiguration" key with the subnets and security groups that you want to include. Bit unfortunate that that much infra information needs to go in code, but I would expect that to work if it's an option.

**Reply by frs171717** [reply]
Thanks, @gibsondan -- that worked. Unfortunately, now when I launch my job from Dagster the job fails with `Run timed out due to taking longer than 600 seconds to start.`. On the AWS end, I can see a new EC2 task gets stuck in `Provisioning` state. I am having trouble finding any more logs. 

This is all quite new to me, so I could very well be doing something silly. My guess was I need to be more careful in selecting my subnets/security groups or maybe something with IAM roles, but perhaps you have another idea?

**Reply by frs171717** [reply]
Ahh, I was indeed doing a number of silly things. After a lot of ECS learning and tinkering I now successfully have an ECS EC2 capacity provider that sends tasks to an auto scaling group with a warm pool.

Unfortunately, I still am almost always getting `Run timed out due to taking longer than 600 seconds to start` errors from Dagster. Monitoring the ECS Agent logs on the EC2 I can see the task is just being set to RUNNING around the 600-700 second mark (hence why on occasion I have a job run successfully.

I need to work on the biggest time-sucks:
1. my very large Docker image pulling slowly anywhere (5 minutes locally) and extremely slowly on my EC2 (8 minutes)
2. my task spending almost 3 minutes in PROVISIONING state before it asks my auto-scaling group to scale from zero to one instances (I am using ECS Managed Scaling, so I suspect this has to do with the CloudWatch alarm they set up for me).


However, @gibsondan, even if I solve these issues it would be great to not need to start in 600 seconds. I believe the setting I need to tweak is `run_monitoring.start_timeout_seconds`, however, I am confused as Dagster Cloud claims the default is already 1200. Perhaps it has a different default in Branch deployments like the one I am testing in? Any advice on increasing this limit in a Branch Deployment would be appreciated, it seems editing deployment settings for Branch Deployments from the UI is not an option.

**Reply by gibsondan** [reply]
That looks like an oversight on our end. We’ll make a change this week that makes branch deployments use the same setting as what you have set for your full deployment in your deployment settings.

**Reply by gibsondan** [reply]
That's now been resolved - branch deployments should be using the same startup timeout as your full deployment now.

**Comment by frs171717** [comment]
Thanks for the assist here, @gibsondan!
